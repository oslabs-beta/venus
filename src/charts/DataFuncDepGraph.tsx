// this is the kind of type of the tree node
interface IByMethod {
    GET?: IService;
    POST?: IService;
    PUT?: IService;
    PATCH?: IService;
}

interface IService {
    service?:string;
    status: string;
    load: string;
    response_time: number;
    error: number;
    availability: number;
    byMethod?: IByMethod

}

interface IDataNode {
    services: IService[];
  }

const test: IDataNode = {
    services: [
        {
            service: 'gogle api',
            status: 'good',
            load: '1.00 hpm',
            response_time:1266,
            error: 0,
            availability: 100,
            byMethod: {
                GET: {
                    status: 'good',
                    load: '0.33 hpm',
                    response_time: 1200,
                    error: 0,
                    availability:100
                },
                POST: {
                  status: 'bad',
                  load: '0.33 hpm',
                  response_time: 1200,
                  error: 0,
                  availability:100
              },
              PATCH: {
                status: 'fair',
                load: '0.33 hpm',
                response_time: 1200,
                error: 0,
                availability:100
            }
            }
        },
        {
            service: 'wizard api',
            status: 'fair',
            load: '1.00 hpm',
            response_time:1266,
            error: 0,
            availability: 100,
            byMethod: {
                GET: {
                    status: 'bad',
                    load: '0.33 hpm',
                    response_time: 1200,
                    error: 0,
                    availability:100
                },
                POST: {
                    status: 'bad',
                    load: '0.33 hpm',
                    response_time: 1200,
                    error: 0,
                    availability:100
                },
            }
        },
        {
          service: 'Solarwinids api',
          status: 'good',
          load: '1.00 hpm',
          response_time:1266,
          error: 0,
          availability: 100,
          byMethod: {
              GET: {
                  status: 'good',
                  load: '0.33 hpm',
                  response_time: 1200,
                  error: 0,
                  availability:100
              },
              POST: {
                  status: 'good',
                  load: '0.33 hpm',
                  response_time: 1200,
                  error: 0,
                  availability:100
              },
          }
      }
    ]
}

/* ----------- CHANGE CHILD ARRAY ---------- */

function changeChildArr(arr: IService[]): TreeNode[]{
    // make a new array with map
    // we will be putting the services inside the array
    const newArr = arr.map( (obj: any) => {
        // declare a variable for the req method objects
        const objReqs = []
        // create request types objs to be put in the req methods array for each endpoint obj
        for (let key in obj.byMethod){
            const reqType: TreeNode = {
            // declare the service type to a service key
            service: key,
            // declare the status type to the status of that req method
            // I purposely declared the obj as "any" type so I could turn the key to a string here
            status: obj.byMethod[key].status}
            // push to req methods array for each endpoint obj
            objReqs.push(reqType)
        }
        // the new Obj for each service w/ type
        const newObj: TreeNode = {
            service: obj.service,
            status: obj.status
           }

 /* -----------THIS IS WHERE YOU WOULD CHANGE "children" to "methods" ------- */
 
         // put the requests in as a "children" key
         
        newObj.children = objReqs
        // return the object to the array
        return newObj
    })
    // return the array
    return newArr
}

// console.log('change arr', changeChildArr(test.services))
let children = changeChildArr(test.services)[0]
console.log('children', children)
// console.log(test.services);

/* ----------- CHANGE MAIN ARRAY --------- */
// this is the kind of type of the tree node
interface TreeNode {
    service: string;
    status: string;
    isExpanded?: boolean;
    children?: TreeNode[];
  }

function changeData(obj: IDataNode): TreeNode{
    let result: TreeNode = {
    // create the main node and its children using helper function
            service: "Services",
            status: "good",
            children: changeChildArr(obj.services)
          }
    return result
    }

const treeData = changeData(test)

// ------ previous dummy data ------- //

// const treeData: TreeNode = {
//     service: "CodeSmith",
//     status: "good",
//     children: [
//       {
//         service: "Google API",
//         status: "good",
//         children: [
//           { service: "GET",
//           status: "bad"},
//           { service: "POST",
//           status: "good" },
//           { service: "PUT",
//           status: "fair" },
//           { service: "DEL",
//           status: "good" },
//           ]
//       },
//       {
//         service: "Plaid API",
//         status: "good",
//         children: [
//           { service: "PUT",
//           status: "good" },
//           { service: "DEL",
//           status: "good"},
//           ]
//       },
//       {
//         service: "Solarwinds API",
//         status: "fair",
//         children: [
//           { service: "GET",
//           status: "good" }
//           ]
//       },
//       {
//         service: "Surfline API",
//           status: "fair",
//         children: [
//           { service: "GET",
//           status: "bad" },
//           { service: "POST",
//           status: "fair" },
//           { service: "PUT",
//           status: "good" },
//           { service: "DEL",
//           status: "fair" },
//           ]
//       },
//       {
//         service: "Yelp API",
//         status: "bad",
//         children: [
//           { service: "GET",
//           status: "bad" },
//           { service: "POST",
//           status: "good" },
//           { service: "PUT",
//           status: "good" },
//           { service: "DEL",
//           status: "good" },
//           ]
//       },
//     ]
//   };

  export { changeChildArr, changeData, test, treeData, IDataNode, IService, TreeNode}