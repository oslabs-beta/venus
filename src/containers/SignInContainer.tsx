import React, { useContext, useState } from "react";
// import data from '../../session_storage/storage.json'
import { globalContext } from '../contexts/globalContext' 
import Form from 'antd/es/form';
import Input from 'antd/es/input';
import Button from 'antd/es/button';
import Card from 'antd/es/card';
import Typography from "antd/es/typography";
import authApi from '../other-modules/authApi';

const { Title } = Typography;


function SignIn():JSX.Element {
  const { verification, setVerification, setServerAddress, } = useContext(globalContext)
  
  /**
   * Deconstruct Form values (serverAddress, secret) 
   * invoke authApi.login (POST request), passing in serverAddress and secret 
   * as key:values in the 'body' object
   * Upon receiving a request, if status is 200 -> authentication was successful
   * Store accessToken in locakStorage, serverAddress in state and 
   * set verification boolean to true
   */
   const onFinish = async (values: any) => {
    const { serverAddress, secret } = values;
    const res = await authApi.login({ serverAddress, secret });
    if (res.status === 200) {
      const { accessToken, refreshToken } = res.data;
      localStorage.setItem('accessToken', accessToken)
      localStorage.setItem('refreshToken', refreshToken)
      setServerAddress(values.serverAddress)
      setVerification(true)
    }
  }
  
  const onFinishFailed = (errorInfo: any) => {
    console.log('Failed:', errorInfo);
  };
  const layout = {
    labelCol: { span: 8 },
    wrapperCol: { span: 16 },
  };
  const tailLayout = {
    wrapperCol: { offset: 8, span: 16 },
  };

  return (
    <div className="loginContainer">
      <Card 
        data-testid="signin-card"
        style={{height: 'fit-content', width: 600, textAlign: 'center'}}
        bordered={true}
        title={<Title>VENUS</Title>}
        hoverable={true}
        >
        <Form
          {...layout}
          data-testid="Login-form"
          name="Login"
          initialValues={{ remember: true }}
          onFinish={onFinish}
          onFinishFailed={onFinishFailed}>
          <Form.Item
            data-testid="serverIP-form"
            label="Server Address"
            name="serverIP"
            rules={[{ required: true, message: 'Please enter valid Server Address.' }]}
          >
            <Input placeholder="Enter Server Address"/>
          </Form.Item>
          <Form.Item
            data-testid="secret-form"
            label="Secret"
            name="secret"
            rules={[{ required: true, message: 'Please enter valid Secret.' }]}
          >
            <Input.Password 
              data-testid="signin-secret"
              placeholder="Enter Secret" 
            />
          </Form.Item>

          <Form.Item {...tailLayout}>
            <Button type="primary" htmlType="submit" data-testid="signin-button">
              Sign In
            </Button>
        </Form.Item>
        </Form>
      </Card>
    </div>
  )
}

export { SignIn };


