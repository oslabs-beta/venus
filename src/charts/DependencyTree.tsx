
import React, { useState } from "react";
import { Group } from "@visx/group";
import { hierarchy, Tree } from "@visx/hierarchy";
import { LinearGradient } from "@visx/gradient";
import { pointRadial } from "d3-shape";
import { AggregateStats } from '../components/AggregateStats';
import { changeChildArr, changeData, treeData} from './DataFuncDepGraph'
import { Select } from "antd";

// import useForceUpdate from "./useForceUpdate";
import {
  LinkHorizontal,
  LinkVertical,
  LinkRadial,
  LinkHorizontalStep,
  LinkVerticalStep,
  LinkRadialStep,
  LinkHorizontalCurve,
  LinkVerticalCurve,
  LinkRadialCurve,
  LinkHorizontalLine,
  LinkVerticalLine,
  LinkRadialLine
} from "@visx/shape";

const { Option } = Select;

const controlStyles = { fontSize: 28 };
export type Props = {
  layout: string;
  orientation: string;
  linkType: string;
  stepPercent: number;
  setLayout: (layout: string) => void;
  setOrientation: (orientation: string) => void;
  setLinkType: (linkType: string) => void;
  setStepPercent: (percent: number) => void;
};

function LinkControls({
  layout,
  orientation,
  linkType,
  stepPercent,
  setLayout,
  setOrientation,
  setLinkType,
  setStepPercent
}: Props) {
  function handleChangeLayout(value: string) {
    setLayout(value);
  }
  function LayoutSelect(): any {
    return (
      <Select
        defaultValue={layout}
        style={{ width: 140 }}
        onChange={handleChangeLayout}
      >
        <Option value="polar">Polar</Option>
        <Option value="cartesian">Cartesian</Option>
      </Select>
    );
  }

  function handleChangeOrientation(value: string) {
    setOrientation(value);
  }
  function OrientationSelect(): any {
    return (
      <Select
        defaultValue={orientation}
        style={{ width: 140 }}
        onChange={handleChangeOrientation}
      >
        <Option value="horizontal">Horizontal</Option>
        <Option value="vertical">Vertical</Option>
      </Select>
    );
  }

  function handleChangeLinkType(value: string) {
    setLinkType(value);
  }
  function LinkTypeSelect(): any {
    return (
      <Select
        defaultValue={linkType}
        style={{ width: 140 }}
        onChange={handleChangeLinkType}
      >
        <Option value="diagonal">Diagonal</Option>
        <Option value="step">Step</Option>
        <Option value="curve">Curve</Option>
        <Option value="line">Line</Option>
      </Select>
    );
  }
  return (
    <div>
      {/* <div width={totalWidth+200}></div> */}
      <label>layout: </label>&nbsp;
      <LayoutSelect width={200}/>
      &nbsp;&nbsp;
      <label>Orientation: </label>&nbsp;
      <OrientationSelect />
      &nbsp;&nbsp;
      <label>Link: </label>&nbsp;
      <LinkTypeSelect />
      {linkType === "step" && layout !== "polar" && (
        <>
          &nbsp;&nbsp;
          {/* <label>step:</label>&nbsp;
          <input
            onClick={(e) => e.stopPropagation()}
            type="range"
            min={0}
            max={1}
            step={0.1}
            onChange={(e) => setStepPercent(Number(e.target.value))}
            value={stepPercent}
            disabled={linkType !== "step" || layout === "polar"}
          /> */}
        </>
      )}
      <br />
    </div>
  );
}

// function for Link Component w/ type assigned properties

function getLinkComponent({
  layout,
  linkType,
  orientation
}: {
  layout: string;
  linkType: string;
  orientation: string;
}): React.ComponentType<any> {
  let LinkComponent: React.ComponentType<any>;

  if (layout === "polar") {
    if (linkType === "step") {
      LinkComponent = LinkRadialStep;
    } else if (linkType === "curve") {
      LinkComponent = LinkRadialCurve;
    } else if (linkType === "line") {
      LinkComponent = LinkRadialLine;
    } else {
      LinkComponent = LinkRadial;
    }
  } else if (orientation === "vertical") {
    if (linkType === "step") {
      LinkComponent = LinkVerticalStep;
    } else if (linkType === "curve") {
      LinkComponent = LinkVerticalCurve;
    } else if (linkType === "line") {
      LinkComponent = LinkVerticalLine;
    } else {
      LinkComponent = LinkVertical;
    }
  } else if (linkType === "step") {
    LinkComponent = LinkHorizontalStep;
  } else if (linkType === "curve") {
    LinkComponent = LinkHorizontalCurve;
  } else if (linkType === "line") {
    LinkComponent = LinkHorizontalLine;
  } else {
    LinkComponent = LinkHorizontal;
  }
  return LinkComponent;
}

// margin from chart to the sides 
const defaultMargin = { top: 110, left: 110, right: 110, bottom: 110 };
// link props type
export type LinkTypesProps = {
  width: number;
  height: number;
  margin?: { top: number; right: number; bottom: number; left: number };
};

// renders the graph
function DependencyGraph({
  width: totalWidth,
  height: totalHeight,
  margin = defaultMargin
}: LinkTypesProps) {
  // set default mode
  const [layout, setLayout] = useState<string>("polar");
  const [orientation, setOrientation] = useState<string>("horizontal");
  const [linkType, setLinkType] = useState<string>("step");
  const [stepPercent, setStepPercent] = useState<number>(0.5);
  const forceUpdate = useForceUpdate();

  const innerWidth = totalWidth - margin.left - margin.right;
  const innerHeight = totalHeight - margin.top - margin.bottom;

  let origin: { x: number; y: number };
  let sizeWidth: number;
  let sizeHeight: number;
// sets chart point origin
  if (layout === "polar") {
    // modifies margin for "polar" layout
    margin = { top: 180, left: 110, right: 110, bottom: 110 }
    origin = {
      x: innerWidth / 2,
      y: innerHeight / 4
    };
    sizeWidth = 2 * Math.PI;
    // can set how spread out chart is here
    sizeHeight = Math.min(innerWidth, innerHeight) / 1.5;
  } else {
    origin = { x: 0, y: 0 };
    if (orientation === "vertical") {
      sizeWidth = innerWidth;
      sizeHeight = innerHeight;
    } else {
      sizeWidth = innerHeight;
      sizeHeight = innerWidth;
    }
  }

  const LinkComponent = getLinkComponent({ layout, linkType, orientation });
// can modify components here i.e. color, stroke size
  return totalWidth < 10 ? null : (
    <div>
      <br />
      <LinkControls
        layout={layout}
        orientation={orientation}
        linkType={linkType}
        stepPercent={stepPercent}
        setLayout={setLayout}
        setOrientation={setOrientation}
        setLinkType={setLinkType}
        setStepPercent={setStepPercent}
      />
      <br/>
      <svg width={600} height={600}>
        <LinearGradient id="links-gradient" from="#fd9b93" to="#fe6e9e" />
        // can change rectangle color
        <rect width={totalWidth+200} height={totalHeight} rx={14} fill="#f5f5f5" />
        <Group top={margin.top} left={margin.left}>
          <Tree
      /*------- Put our data variable in place of treeData --------*/
            root={hierarchy(treeData, (d) => (d.isExpanded ? null : d.children))}
            size={[sizeWidth, sizeHeight]}
            separation={(a, b) => (a.parent === b.parent ? .4 : .4) / a.depth}
          >
            {(tree) => (
              <Group top={origin.y} left={origin.x}>
                {tree.links().map((link, i) => (
                  <LinkComponent
                    key={i}
                    data={link}
                    percent={stepPercent}
                    stroke="#03c0dc"
                    strokeWidth="1.75"
                    fill="none"
                  />
                ))}
              // settings for children of tree graph
                {tree.descendants().map((node, key) => {
                  const width = 160;
                  const height = 30;

                  let top: number;
                  let left: number;
                  // if (node.data.status === 'good'){ console.log('node', node)}
                  if (layout === "polar") {
                    const [radialX, radialY] = pointRadial(node.x, node.y);
                    top = radialY;
                    left = radialX;
                  } else if (orientation === "vertical") {
                    top = node.y;
                    left = node.x;
                  } else {
                    top = node.x;
                    left = node.y;
                  }
                  // THIS LINE OF CODE IS THE EQUATION FOR CHANGING COLOR BASED ON NODE HEALTH
                  const colorChange = (node.data.status === 'good') ? "#F6FFED" : (node.data.status === 'fair') ? '#FFF7E6' : (node.data.status === 'bad') ? '#FFF1F0' : (node.data.status)
                  const changeChildren = (node.data.children) ? colorChange : colorChange
                  
                  const colorChangeText = (node.data.status === 'good') ? "#52C41A" : (node.data.status === 'fair') ? '#FA8C16' : (node.data.status === 'bad') ? '#F5222D' : (node.data.status)
                  const changeChildrenText = (node.data.children) ? colorChangeText : colorChangeText
                  
                  const colorChangeBorder = (node.data.status === 'good') ? "#B7EB8F" : (node.data.status === 'fair') ? '#FFD591' : (node.data.status === 'bad') ? '#FFA39E' : (node.data.status)
                  const changeChildrenBorder = (node.data.children) ? colorChangeBorder : colorChangeBorder
                  

                  return (
                    <Group top={top} left={left} key={key}>
                      {node.depth === 0 && (
                        <circle
                          r={12}
                          fill="#f5f5f5"
                          onClick={() => {
                            node.data.isExpanded = !node.data.isExpanded;
                            console.log(node);
                            forceUpdate();
                          }}
                        />
                      )}
                      {node.depth !== 0 && (
                        <rect
                          height={height}
                          width={node.data.children ? "24%" : "10%"}
                          y={-height / 2}
                          x={node.data.children ? -(width-18) / 2 : -(width-100) / 2}
                          // fill of individual node boxes
                          fill={changeChildren}
                          // change border here ------------
                          stroke= {colorChangeBorder}
                          strokeWidth={1.75}
                          strokeDasharray={node.data.children ? "0" : "2,2"}
                          strokeOpacity={node.data.children ? 1 : 0.6}
                          rx={node.data.children ? 0 : 10}
                          onClick={() => {
                            node.data.isExpanded = !node.data.isExpanded;
                            console.log(node);
                            forceUpdate();
                          }}
                        />
                      )}
                      <text
                        dy=".33em"
                        fontSize={20}
                        fontFamily= 'Roboto'
                        textAnchor="middle"
                        style={{ pointerEvents: "none" }}
                        fill={colorChangeText}
                      >
                        {node.data.service}
                      </text>
                    </Group>
                  );
                })}
              </Group>
            )}
          </Tree>
        </Group>
      </svg>
    </div>
  );
}

function useForceUpdate() {
  const [value, setValue] = useState<number>(0);
  return () => setValue(value => value + 1); // update state to force render
}

export { DependencyGraph, treeData, getLinkComponent, LinkControls };
