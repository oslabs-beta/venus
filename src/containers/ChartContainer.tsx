  /**
 * @name ChartContainer
 * @desc Child of Dashboard, Parent container that holds and displays each Chart
 */
import React, { useContext, useEffect, useState} from "react";
import { CardDropDown } from "../components/CardDropDown";
import { historicalContext } from "../contexts/historicalContext"
import { AggregateStats } from "../components/AggregateStats";
import { Availability } from "../charts/AvailabilityChart";
import { LoadChart } from "../charts/LoadChart";
import Row from 'antd/es/row';
import Col from 'antd/es/col';
import Card from 'antd/es/card';
import Divider from "antd/es/divider";
import Title from "antd/es/typography/Title";
import Radio from 'antd/es/radio'
import Button from 'antd/es/button'
import Skeleton from 'antd/es/skeleton'

function ChartContainer(): JSX.Element {
  const [bool, setBool] = useState(false)
  const { aggregate, setAggregate, service, serviceData, setService, setServiceData } = useContext(historicalContext)
  
  const test: any = {
    status: 0,
    load: 12,
    response_time: 12, 
    error: 12,
    availability: 12,
  }
  let services:any = [
    'service a', 'service b', 'service c'
  ]
  
  useEffect(() => {
    setAggregate(test)
    
    // fetch('localhost:3000/get/historicdata').then(
    //   response => response.json()
    // ).then(
    //   data => console.log(data)
    // )
    console.log('rerender')
    fetch('https://gw.alipayobjects.com/os/bmw-prod/55424a73-7cb8-4f79-b60d-3ab627ac5698.json')
      .then((response) => response.json())
      .then((json) => { 
        setServiceData(json)
        setBool(true)
      })
      .catch((error) => {
        console.log('fetch data failed', error);
      });
  },[])

  
// temporal options for chart displays
  const options = [
    { label: 'Last Hour', value: 'lastHour' },
    { label: 'Last Day', value: 'lastDay' },
    { label: 'Last Week', value: 'lastWeek' },
    { label: 'Last Month', value: 'lastMonth' },
  ];

// select data range to display from historical state
  const filterData = (e:any) => {
    console.log('radio checked', e.target.value, service);
  };

  const refreshData = () => {
    // fetch('localhost:3000/get/historicdata').then(
    //   response => response.json()
    // ).then(
    //   data => console.log(data)
    //   setServiceData(data)
    // )
  };
  
  if(!bool){
     return <div>LOADING</div>
  } else {
  return (
    <div id="chartContainer">
      <AggregateStats
      error={aggregate.error}
      response_time={aggregate.response_time}
      load={aggregate.load}
      availability={aggregate.availability}
      />
      <Divider><Title level={3}>Historical Status</Title></Divider>
      <div className="rangeSelectorContainer">
        <CardDropDown services={services}  />
        
        <Radio.Group style={{marginLeft: '10px'}} optionType="button"  onChange={filterData} options={options}/>

        <Button type="primary" style={{marginLeft: '10px'}} onClick={refreshData}>Refresh Data</Button>
      </div>
        <Row gutter={[32,32]}>
          <Col span={12}>
            <Title level={5}>Availability</Title>
            <Card bordered={true} hoverable={true} style={{width: "500px"}}>
              <Availability />
            </Card>
          </Col>
          <Col span={12}>
            <Title level={5}>Latency</Title>
            <Card bordered={true} hoverable={true} style={{width: "500px"}}>
              <Availability />
            </Card>
          </Col>
        </Row>
        <Row gutter={[32,32]}>
          <Col span={12}>
            <Title level={5}>Error Rate</Title>
            <Card bordered={true} hoverable={true} style={{width: "500px"}}>
              <LoadChart />
            </Card>
          </Col>
          <Col span={12}>
          <Title level={5}>Load</Title>
            <Card bordered={true} hoverable={true} style={{width: "500px"}}>
              <LoadChart />
            </Card>
          </Col>
        </Row>
    </div>
  )
}
}
export { ChartContainer, Availability, LoadChart, Button, AggregateStats }