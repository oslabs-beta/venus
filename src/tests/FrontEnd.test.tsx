/**
 * @jest-environment jsdom
 */

import React from 'react';
import { changeChildArr, changeData, IDataNode, test, IService, TreeNode} from "../charts/DataFuncDepGraph"
import { DependencyGraph,treeData, getLinkComponent, LinkControls  } from '../charts/DependencyTree'
import { SettingsContainer } from '../containers/ServiceSettingsContainer';
import { render } from "@testing-library/react";
import { SignIn } from "../containers/SignInContainer";
import { ChartContainer,  Availability, LoadChart, AggregateStats} from "../containers/ChartContainer";
import { Dashboard } from "../containers/Dashboard";

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// describes what the test is
describe('jest testing is in progress', () => {
  // what the specific test is
    it('jest testing activated', () => {
      // what is expected
      expect(true).toBeTruthy();
    })
  }) 

describe('jest snapshot testing', () => {
  describe('dependency graph snapshot testing', () => {
    it('the dependency chart component matches the snapshot', () => {
      expect(DependencyGraph).toMatchSnapshot();
    })

    it('the getLinkComponent component matches the snapshot', () => {
        expect(getLinkComponent).toMatchSnapshot();
    })

    it('the LinkControls component matches the snapshot', () => {
      expect(LinkControls).toMatchSnapshot();
    })
  })

  describe('dummy data snapshot testing', () => {
    it('the dummy data matches the snapshot', () => {
      expect(test).toMatchSnapshot();
    })
    
    it('the dummy data matches the snapshot', () => {
      expect(treeData).toMatchSnapshot();
    })
  })
})

describe('jest data parser testing', () => {
  it('dummy data ran through parser correctly matches post parser chart data', () => {
    expect(changeData(test)).toEqual(treeData)
  })

  describe('jest data child array parser testing', () => {
    it('check if first element from result array exists', () => {
      expect(changeChildArr(test.services)[0]).toBeTruthy();
    })

    it('check if result array is an object', () => {
      expect(typeof changeChildArr(test.services)).toBe('object');
    })
 
    it('check if first element from result array is an object', () => {
      expect(typeof changeChildArr(test.services)[0]).toBe('object');
    })

    it('check if first element from result array has "service" property', () => {
      expect(changeChildArr(test.services)[0]).toHaveProperty('service')
    })

    it('check if first element from result array has "status" property', () => {
      expect(changeChildArr(test.services)[0]).toHaveProperty('status')
    })

    it('check if first element from result array has "children" property', () => {
      expect(changeChildArr(test.services)[0]).toHaveProperty('children')
    })

    it('check if dummy data result array from child parser has length of 2', () => {
      console.log(changeChildArr(test.services))
      expect(changeChildArr(test.services)).toHaveLength(3)
    })
  })
})

/** why we are not using interface in our jest testing file
 * https://stackoverflow.com/questions/48813882/testing-typescript-interface-with-jest
 */

//--------Ground work for mock testing---------//

// describe('mock test with dummy data', () => {
//   const mockCallback = jest.fn(x => changeChildArr(x));
  // function testChildFunc(items: IService[], callback: (arr: IService[]) => TreeNode[]){
//       callback(items);
//   }
//   testChildFunc(test.services,mockCallback);
//   it('mock test for changeChildArr with dummy data', () => {
//     // The mock function is called twice
//     // expect(mockCallback.mock.calls.length).toBe(1);

//     // The first argument of the first call to the function was 0
//     expect(mockCallback.mock.calls[0][0]).toBe(changeChildArr(test.services)[0]);
//   })

// })

//----- ground work for testing endpoints with promise testing with chai-as-promised -----//

// const chai = require("chai");
// const chaiAsPromised = require("chai-as-promised");
// chai.use(chaiAsPromised);
// const expect = chai.expect

// it('Promise should pass', function() {
//   return expect(readAndWriteToDB()).to.be.fulfilled
// })

//----- ground work for enzyme testing -----//
import { shallow } from 'enzyme';
// import {App} from '../app'; // Introduce the corresponding React component
describe('Enzyme testing is in progress', () => {
  it('Enzyme testing activated', () => {
    expect(true).toBeTruthy();
  });
})

// -------- testing sign in container --------- //

// describe('enzyme test for react', () => {
//   it(('checking the sign in', () => {
//   const onCountChange = jest.fn();
//   let wrapper;
//   })
//   beforeEach(() => {
//     wrapper = mount(SignIn())
//   })
// })

  describe('checking sign in container', () => {
    it('sign in container exists', () => {
      const signInContainer:any = shallow(<SignIn/>)
      expect(signInContainer).toBeTruthy();
    });

    it('expects secret form to be in SignIn conatainer', () => {
      const { queryByTestId } = render(<SignIn />);
      const addSignInButton = queryByTestId("secret-form");
      expect(addSignInButton).toBeInTheDocument();
    })

    it('expects server IP form to be in SignIn container', () => {
      const { queryByTestId } = render(<SignIn />);
      const addSignInButton = queryByTestId("serverIP-form");
      expect(addSignInButton).toBeInTheDocument();
    })

    it('expects button to be in SignIn container', () => {
      const { queryByTestId } = render(<SignIn />);
      const addSignInButton = queryByTestId("signin-button");
      expect(addSignInButton).toBeInTheDocument();
    })   
    
    it('expects signin card to be in SignIn container', () => {
      const { queryByTestId } = render(<SignIn />);
      const addSignInButton = queryByTestId("signin-card");
      expect(addSignInButton).toBeInTheDocument();
    })     

    it('expects "Enter Secret" component to be in SignIn container', () => {
      const { queryByTestId } = render(<SignIn />);
      const addSignInButton = queryByTestId("signin-secret");
      expect(addSignInButton).toBeInTheDocument();
    })     

  })

// -------- testing chart container with enzyme (to bring in: react hooks with enzyme for chart container/MainDisplay)--------- //

describe('testing for chart container', () => {
  it('chart container exists', () => {
    const chartContainer:any = shallow(<ChartContainer />)
    expect(chartContainer).toBeTruthy();
  });

  it('the chart container container matches the snapshot', () => {
    expect(ChartContainer).toMatchSnapshot();
  })

  it('availability exists in the chart container', () => {
    const availabilityComponent:any = shallow(<Availability />)
    expect(availabilityComponent).toBeTruthy();
  });

  it('load chart exists in the chart container', () => {
    const loadChartComponent:any = shallow(<LoadChart />)
    expect(loadChartComponent).toBeTruthy();
  });

  it('aggregate stats exist in the chart container', () => {
    const aggregateStatsComponent:any = shallow(<AggregateStats />)
    expect(aggregateStatsComponent).toBeTruthy();
  });
})

// ----- checking main Display ----- //

// describe('testing for maindisplay container', () => {
//   it('main display container exists', () => {
//     const mainDisplayContainer:any = render(<MainDisplay />)
//     expect(mainDisplayContainer).toBeTruthy();
//   });
// })

// ----- checking Dashboard container ----- //

// describe('testing for maindisplay container', () => {
//   it('main display container exists', () => {
//     const dashboardContainer:any = render(<Dashboard />)
//     expect(dashboardContainer).toBeTruthy();
//   });
// })

// ----- checking service settings container ------ //

// describe('checking service settings container', () => {
//   it('service settings container exists', () => {
//     const serviceSettingsContainer:any = render(<SettingsContainer/>)
//     expect(serviceSettingsContainer).toBeTruthy();
//   })
//   // it('expects button to be in SignIn component', () => {
//   //   const { queryByTestId } = render(<SettingsContainer />);
//   //   const addReactComponent = queryById("dashboard");
//   //   expect(addReactComponent).toBeInTheDocument();
//   // })
// })
