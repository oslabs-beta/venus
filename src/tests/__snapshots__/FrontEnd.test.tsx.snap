// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest snapshot testing dependency graph snapshot testing the LinkControls component matches the snapshot 1`] = `[Function]`;

exports[`jest snapshot testing dependency graph snapshot testing the dependency chart component matches the snapshot 1`] = `[Function]`;

exports[`jest snapshot testing dependency graph snapshot testing the getLinkComponent component matches the snapshot 1`] = `[Function]`;

exports[`jest snapshot testing dummy data snapshot testing the dummy data matches the snapshot 1`] = `
Object {
  "services": Array [
    Object {
      "availability": 100,
      "byMethod": Object {
        "GET": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "good",
        },
        "PATCH": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "fair",
        },
        "POST": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "bad",
        },
      },
      "error": 0,
      "load": "1.00 hpm",
      "response_time": 1266,
      "service": "gogle api",
      "status": "good",
    },
    Object {
      "availability": 100,
      "byMethod": Object {
        "GET": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "bad",
        },
        "POST": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "bad",
        },
      },
      "error": 0,
      "load": "1.00 hpm",
      "response_time": 1266,
      "service": "wizard api",
      "status": "fair",
    },
    Object {
      "availability": 100,
      "byMethod": Object {
        "GET": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "good",
        },
        "POST": Object {
          "availability": 100,
          "error": 0,
          "load": "0.33 hpm",
          "response_time": 1200,
          "status": "good",
        },
      },
      "error": 0,
      "load": "1.00 hpm",
      "response_time": 1266,
      "service": "Solarwinids api",
      "status": "good",
    },
  ],
}
`;

exports[`jest snapshot testing dummy data snapshot testing the dummy data matches the snapshot 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "service": "GET",
          "status": "good",
        },
        Object {
          "service": "POST",
          "status": "bad",
        },
        Object {
          "service": "PATCH",
          "status": "fair",
        },
      ],
      "service": "gogle api",
      "status": "good",
    },
    Object {
      "children": Array [
        Object {
          "service": "GET",
          "status": "bad",
        },
        Object {
          "service": "POST",
          "status": "bad",
        },
      ],
      "service": "wizard api",
      "status": "fair",
    },
    Object {
      "children": Array [
        Object {
          "service": "GET",
          "status": "good",
        },
        Object {
          "service": "POST",
          "status": "good",
        },
      ],
      "service": "Solarwinids api",
      "status": "good",
    },
  ],
  "service": "Services",
  "status": "good",
}
`;

exports[`testing for chart container the chart container container matches the snapshot 1`] = `[Function]`;
